### PR TITLE
Let postgresql-simple make the transaction

### PR DESCRIPTION
**Pull request by Ericson2314**
We want to be in a serializable transaction that retries.

### CODE REVIEW

**Comment by 3noch** [comment]
*This* is the implementation I was hoping for. Awesome. However @danbornside had the idea that we should specify the retry strategy more explicitly or at least have that as an option.

**Comment by 3noch** [comment]
IMO the ideal solution would be to avoid MonadBaseControl altogether and re-implement a lifted version of `withTransactionSerializable` on top of `genericBracket` from `MonadMask`.

**Comment by 3noch** [comment]
This of course changes more things about groundhog though.

**Comment by 3noch** [comment]
Merged in `develop` and fixed conflict.

File: backend-db/Rhyolite/Backend/DB.hs [hunk 1]
```diff
@@ -66,20 +68,35 @@ type Db m = (PersistBackend m, PostgresRaw m, SqlDb (PhantomDb m))
...
-  runDb (Identity db) = withResource db . runDbConn . withSchema (SchemaName "public")
+  runDb (Identity db) = runDb $ WithSchema (SchemaName "public") db
 
 -- | Runs a database action in a specific schema
 instance RunDb WithSchema where
-  runDb (WithSchema schema db) = withResource db . runDbConn . withSchema schema
+  runDb (WithSchema schema db) = withResource db
+    . runDbConnNoTransaction
+    . (\(DbPersist (ReaderT f)) -> DbPersist $ ReaderT $ \conn@(Postgresql conn') ->
+          liftBaseThrough (withTransactionSerializable conn') $ f conn)
```
Comment (ali-abrar) [inline backend-db/Rhyolite/Backend/DB.hs]: this looks a bit more complex than I was expecting. can you explain a little?

**Comment by 3noch** [comment]
Actually I think we probably don't want this. The automatic retrying makes DbPersist unsound as a monad transformer in general. For example, running on top of State or Writer will cause Bad Things if we retry. And of course liftIO becomes much more dangerous too since it duplicates effects. Also it's possible to mask serialization errors by using catch in the code. This could cause infinite loops.

I think the only solution is to build a monomorphic, non-transformer monad that doesn't have MonadIO and likely not MonadMask either. MonadBaseControl is way out of the question.

**Comment by 3noch** [comment]
After discussing this with in the tech Q&A meeting we all agree that we should go with a simpler, monomorphic monad. I'll submit something.

**Comment by 3noch** [comment]
See #106 for an alternative

**Comment by ali-abrar** [comment]
@3noch can this be closed now?

**Comment by ali-abrar** [comment]
Based on #106 it looks like this can be closed.

**Pull request closed without merging**